fix: Fix write algorithm in delayed init mode

dmga44 opened the pull request:
This PR targets fixing the inconsistent selection of write
algorithm when using the `sfsdelayedinit` option in client.

Main feature of sfsdelayedinit option is to be able to start the mount
without an available master, which means it is not expected after short
notice to know the master version. This means client is not supposed to
display any warning message right after mounting and select the
appropiate write algorithm after knowing the current master version.

Implements the support for the client to switch the
write algorithm used when needed depending on master version.

An integration was added and another one was enhanced in order
to test the proposed changes. Also, a specific test for the
`sfsdelayedinit` option was added in order to check its expected
behavior.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
fix: Fix write algorithm in delayed init mode

This commit targets fixing the inconsistent selection of write
algorithm when using the sfsdelayedinit option in client.

Main feature of sfsdelayedinit option is to be able to start the mount
without an available master, which means it is not expected after short
notice to know the master version. This means client is not supposed to
display any warning message right after mounting and select the
appropiate write algorithm after knowing the current master version.

This commit also implements the support for the client to switch the
write algorithm used when needed depending on master version.

Signed-off-by: Dave <dave@leil.io>

dmga44 commented on src/mount/mastercomm.cc:
```diff
@@ -196,6 +200,70 @@ struct InitParams {
+
+		assert(gUseInodeBasedWriteAlgorithm && gIsInodeBasedWriteAlgorithmInitialized);
+		write_data_term();
+		gIsChunkBasedWriteAlgorithmInitialized = false;
```
Nice catch.

dmga44 commented on src/mount/mastercomm.cc:
```diff
@@ -196,6 +200,70 @@ struct InitParams {
+	bool useInodeBasedWriteAlgorithm = gSaunaFSInitParams.use_inode_based_write_algorithm;
+
+	// Dangerous case, using chunk based algorithm by params and master does not support it.
+	if (!gSaunaFSInitParams.use_inode_based_write_algorithm &&
+	    masterversion < kFirstVersionWithChunkBasedWriteAlgorithm) {
+		if (isFromMainThread) {
+			fprintf(stderr,
+			        "Metadata server version v%s is too old, using inode based write algorithm"
+			        "(sfsuseinodebasedwritealgorithm=1). "
+			        "Required minimum version for chunk based write algorithm is v%s.\n",
+			        saunafsVersionToString(masterversion).c_str(),
+			        saunafsVersionToString(kFirstVersionWithChunkBasedWriteAlgorithm).c_str());
+		} else {
+			safs::log_warn(
+			    "Metadata server version v{} is too old, using inode based write algorithm"
+			    "(sfsuseinodebasedwritealgorithm=1). "
+			    "Required minimum version for chunk based write algorithm is v{}.",
+			    saunafsVersionToString(masterversion),
+			    saunafsVersionToString(kFirstVersionWithChunkBasedWriteAlgorithm));
+		}
```
This function is only going to be called after a connection with a master has been stablished. Therefore, the master version should not be never 0.

dmga44 committed:
test: Add test for switching write algorithm

This commit improves the test test_upgrade_new_mount_with_old_saunafs
to make sure it writes more than one chunk, which should be previously
triggering the incompatibility issues between new mounts and old
masters.

It also adds a new integration test:
test_upgrade_master_switch_write_algo to make sure old/new clients
could read/write to systems having old/new masters.

Signed-off-by: Dave <dave@leil.io>

uristdwarf commented on src/mount/mastercomm.cc:
```diff
@@ -196,6 +200,68 @@ struct InitParams {
+	    !gSaunaFSInitParams.use_inode_based_write_algorithm &&
+	    masterversion >= kFirstVersionWithChunkBasedWriteAlgorithm) {
+		if (isFromMainThread) {
+			fprintf(stderr,
+			        "New metadata server supports chunk based write algorithm "
+			        "(current master version: %s, required: %s).\n",
+			        saunafsVersionToString(masterversion).c_str(),
+			        saunafsVersionToString(kFirstVersionWithChunkBasedWriteAlgorithm).c_str());
```
Maybe use fmt instead of fprintf? fmt at least does compile time type checking.

dmga44 replied:
Well, I did it just like the rest of the logging of client initialization, the way that it appears like all the other messages.

uristdwarf commented on src/mount/mastercomm.cc:
```diff
@@ -196,6 +200,68 @@ struct InitParams {
+			    saunafsVersionToString(kFirstVersionWithChunkBasedWriteAlgorithm));
+		}
+
+		assert(gUseInodeBasedWriteAlgorithm && gIsInodeBasedWriteAlgorithmInitialized);
```
Rather than just assert (because of ndebug), I would add passert or something to explicitly crash this even in release builds.

dmga44 replied:
Applied.

rolysr commented on src/mount/writedata.cc:
```diff
@@ -683,6 +683,10 @@ void *write_worker(void *) {
                      uint32_t writewindowsize, uint32_t chunkserverTimeout_ms,
                      uint32_t cachePerInodePercentage, uint32_t waveTimeout) {
+	if (gIsInodeBasedWriteAlgorithmInitialized) {
+		return;
```
As a doubt, why is this kind of check needed on the `write_data_init`? On which scenario will the corresponding write algorithm be initialized before the `write_data_init`?

dmga44 replied:
The proposed solution to the issue calls `write_data_init` if connected to a new master, so a previous master connection to a master supporting the previously used write algorithm could have called `write_data_init` before and have the structures initialized before the next call.

rolysr commented on tests/test_suites/ShortSystemTests/test_upgrade_master_switch_write_algo.sh:
```diff
@@ -0,0 +1,85 @@
+# Force the master to create new sessions (and clients to connect instead of reconnect)
+rm $(get_current_master_sessions_file)
+
+# Don't know why, but we to bypass the saunafs_master_daemon function
```
Check this comment. Should it be `but we have to bypass`?

dmga44 replied:
True, thank you.

rolysr commented on tests/test_suites/ShortSystemTests/test_upgrade_master_switch_write_algo.sh:
```diff
@@ -0,0 +1,85 @@
+assert_success generate_file new_mount_old_master_file
+assert_success file-validate new_mount_old_master_file
+
+# Test if all files produced so far are readable on legacy mount:
```
Should this comment here refer to new mount instead of legacy? As I see from this next lines, you are reading legacy-created files on new mount.

dmga44 replied:
Nope.

I'm reading files from legacy mount (`"${info[mount0]}"` implies it) that were created from the new mount.

rolysr reviewed (commented):
Consider checking my suggestions.

uristdwarf commented on tests/test_suites/ShortSystemTests/test_upgrade_master_switch_write_algo.sh:
```diff
@@ -0,0 +1,85 @@
+  MOUNTS=2 \
+  START_WITH_LEGACY_SAUNAFS=YES \
+  USE_RAMDISK=YES \
+  MOUNT_EXTRA_CONFIG="sfscachemode=NEVER" \
```
What about `sfsdelayedinit` option?

dmga44 replied:
Well, this specific test checks that the change of the write algorithm depending on the connected master is working.

The `sfsdelayedinit` option is kind of a subset of this type of behavior (is only to skip the wait at begining when no master is available), while the test checks that new client can connect to old master and use inode based wirte algorithm and then connect to new master and use chunk based write algorithm. I could add a test for the `sfsdelayedinit` option, but it would look very similar to this one.

uristdwarf reviewed (approved):
Looks fine.

rolysr reviewed (approved):
Nice catch. LGTM.

dmga44 committed:
test: Add a specific test for delayed init option

The sfsdelayedinit option was not being tested in any of out tests,
therefore, it would be safe to check its behavior in some test. This
commit adds such test.

Signed-off-by: Dave <dave@leil.io>

ralcolea reviewed (approved):
Great job @dmga44! 👍 
Nice job fixing the Ganesha multi export test!
